### ISSUE TITLE
vmlistener_test tearDown logic is unstable

### ISSUE THREAD

**Issue by shaominchen** [vmware/docker-volume-vsphere#1308]
vmlistener_test tearDown logic is unstable: it fails intermittently when I run e2e test locally:

2017/05/30 18:28:55 Failed to invoke command [docker volume rm vmlistener_test_volume_1496168883]: exit status 1

----------------------------------------------------------------------
FAIL: vmlistener_test.go:61: VMListenerTestParams.TearDownTest

vmlistener_test.go:67:
    c.Assert(err, IsNil, Commentf(out))
... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc420167620), Stderr:[]uint8(nil)} ("exit status 1")
... Error response from daemon: unable to remove volume: remove vmlistener_test_volume_1496168883: VolumeDriver.Remove: Remove failure - volume is still mounted.  volume=vmlistener_test_volume_1496168883, refcount=1

----------------------------------------------------------------------
PANIC: vmlistener_test.go:80: VMListenerTestParams.TestKillVM

... Panic: Fixture has panicked (see related PANIC)
OOPS: 8 passed, 1 FAILED, 1 MISSED
--- FAIL: Test (274.17s)

**Comment by shaominchen** [comment]
@shuklanirdesh82 Can you please take a look?

**Comment by shuklanirdesh82** [comment]
@shaominchen can you please supply the whole test run log?

**Comment by shaominchen** [comment]
@shuklanirdesh82 My local environment has been overwritten by other tests. Will provide if I reproduce this. Meanwhile you can also try to run e2e test locally a few times - it should not be difficult to reproduce.

**Comment by shuklanirdesh82** [comment]
@shaominchen I did multiple times (6 times so far with master code) and not able to reproduce hence the request for full logs.

**Comment by shuklanirdesh82** [comment]
BTW .. PR#1306 should take care of such intermittent failure (supplied a quick PR) until we are getting rid off all `log.fatalf()` messages.

**Comment by shaominchen** [comment]
@shuklanirdesh82 Here are the complete logs:

shchen-m02:docker-volume-vsphere shchen$ make e2e-test
/Library/Developer/CommandLineTools/usr/bin/make --directory=vmdk_plugin e2e-test
Makefile:152: warning: overriding commands for target `clean'
../plugin/Makefile:87: warning: ignoring old commands for target `clean'

=> Running target e2e-test Tue May 30 12:16:14 PDT 2017

../misc/scripts/build.sh e2e-dkrVolDriver-test "10.161.10.254" "10.161.4.212" "10.161.28.209"
Makefile:152: warning: overriding commands for target `clean'
../plugin/Makefile:87: warning: ignoring old commands for target `clean'
make: Entering directory `/go/src/github.com/vmware/docker-volume-vsphere/vmdk_plugin'

=> Running target e2e-dkrVolDriver-test Tue May 30 19:16:14 UTC 2017

GO15VENDOREXPERIMENT=1 go test github.com/vmware/docker-volume-vsphere/tests/e2e -v
=== RUN   Test
2017/05/30 19:16:15 START: basic-test.TestVolumeIsolation
2017/05/30 19:16:15 Creating volume [basic_test_volume_1496171775] on VM [10.161.4.212]
2017/05/30 19:16:18 Checking volume [basic_test_volume_1496171775] availability from VM [10.161.4.212]
2017/05/30 19:16:18 Checking volume [basic_test_volume_1496171775] availability from VM [10.161.28.209]
2017/05/30 19:16:19 Destroying volume [basic_test_volume_1496171775]
2017/05/30 19:16:20 END: basic-test.TestVolumeIsolation
2017/05/30 19:16:20 START: basic-test.TestVolumeLifecycle
2017/05/30 19:16:20 Creating volume [basic_test_volume_1496171780] on VM [10.161.4.212]
2017/05/30 19:16:22 Checking volume [basic_test_volume_1496171780] availability from VM [10.161.4.212]
2017/05/30 19:16:22 Attaching volume [basic_test_volume_1496171780] on VM [10.161.4.212]
2017/05/30 19:16:24 Confirming attached status for volume [basic_test_volume_1496171780]
2017/05/30 19:16:26 Removing container [basic_test_container_1496171780] on VM [10.161.4.212]
2017/05/30 19:16:27 Confirming detached status for volume [basic_test_volume_1496171780]
2017/05/30 19:16:27 Sleep for 2 seconds
2017/05/30 19:16:31 Destroying volume [basic_test_volume_1496171780]
2017/05/30 19:16:32 Checking volume [basic_test_volume_1496171780] availability from VM [10.161.4.212]
2017/05/30 19:16:33 Creating volume [basic_test_volume_1496171780] on VM [10.161.28.209]
2017/05/30 19:16:35 Checking volume [basic_test_volume_1496171780] availability from VM [10.161.28.209]
2017/05/30 19:16:35 Attaching volume [basic_test_volume_1496171780] on VM [10.161.28.209]
2017/05/30 19:16:37 Confirming attached status for volume [basic_test_volume_1496171780]
2017/05/30 19:16:39 Removing container [basic_test_container_1496171780] on VM [10.161.28.209]
2017/05/30 19:16:40 Confirming detached status for volume [basic_test_volume_1496171780]
2017/05/30 19:16:40 Sleep for 2 seconds
2017/05/30 19:16:44 Destroying volume [basic_test_volume_1496171780]
2017/05/30 19:16:44 Checking volume [basic_test_volume_1496171780] availability from VM [10.161.28.209]
2017/05/30 19:16:45 END: basic-test.TestVolumeLifecycle
2017/05/30 19:16:45 Finding VM name from IP Address [10.161.28.209]
2017/05/30 19:16:46 Initializing the SingleNode Config DB on esx [10.161.10.254] 
2017/05/30 19:16:47 START: defaultvmgroup.TestAddVMToDefaultTenant
2017/05/30 19:16:47 Adding VM [VM1.2] to a vmgroup [_DEFAULT ] on esx [10.161.10.254] 
2017/05/30 19:16:48 END: defaultvmgroup.TestAddVMToDefaultTenant
2017/05/30 19:16:48 START: defaultvmgroup.TestRemoveDefaultTenantVMs
2017/05/30 19:16:48 Removing VM [VM1.2] from a vmgroup [_DEFAULT ] on esx [10.161.10.254] 
2017/05/30 19:16:49 END: defaultvmgroup.TestRemoveDefaultTenantVMs
2017/05/30 19:16:49 START: defaultvmgroup.TestReplaceDefaultTenantVMs
2017/05/30 19:16:49 Replacing VM [VM1.2] from a vmgroup [_DEFAULT ] on esx [10.161.10.254] 
2017/05/30 19:16:50 END: defaultvmgroup.TestReplaceDefaultTenantVMs
2017/05/30 19:16:50 Removing the SingleNode Config DB on esx [10.161.10.254] 
2017/05/30 19:16:51 Creating volume [restart_test_volume_1496171811] on VM [10.161.28.209]
2017/05/30 19:16:55 START: restart_test.TestPluginKill
2017/05/30 19:16:55 Attaching volume [restart_test_volume_1496171811] on VM[10.161.28.209]
2017/05/30 19:16:56 Confirming attached status for volume [restart_test_volume_1496171811]
2017/05/30 19:16:58 Killing vDVS plugin on VM [10.161.28.209]
2017/05/30 19:16:59 Sleep for 2 seconds
2017/05/30 19:17:01 Stopping container [restart_test_container_1496171811] on VM [10.161.28.209]
2017/05/30 19:17:13 Starting container [restart_test_container_1496171811] on VM [10.161.28.209]
2017/05/30 19:17:14 Stopping container [restart_test_container_1496171811] on VM [10.161.28.209]
2017/05/30 19:17:25 Confirming detached status for volume [restart_test_volume_1496171811]
2017/05/30 19:17:25 Sleep for 2 seconds
2017/05/30 19:17:29 END: restart_test.TestPluginKill
2017/05/30 19:17:29 Removing container [restart_test_container_1496171811] on VM [10.161.28.209]
2017/05/30 19:17:30 Destroying volume [restart_test_volume_1496171811]
2017/05/30 19:17:30 Creating volume [restart_test_volume_1496171850] on VM [10.161.28.209]
2017/05/30 19:17:33 START: restart_test.TestVolumeStaleMount
2017/05/30 19:17:33 Attaching volume [restart_test_volume_1496171850] on VM [10.161.28.209]
2017/05/30 19:17:34 Confirming attached status for volume [restart_test_volume_1496171850]
2017/05/30 19:17:36 Killing docker on VM [10.161.28.209]
2017/05/30 19:17:36 Sleep for 2 seconds
2017/05/30 19:17:39 Confirming detached status for volume [restart_test_volume_1496171850]
2017/05/30 19:17:39 Sleep for 2 seconds
2017/05/30 19:17:56 Sleep for 2 seconds
2017/05/30 19:17:58 Sleep for 2 seconds
2017/05/30 19:18:02 END: restart_test.TestVolumeStaleMount
2017/05/30 19:18:02 Removing container [restart_test_container_1496171850] on VM [10.161.28.209]
2017/05/30 19:18:03 Destroying volume [restart_test_volume_1496171850]
2017/05/30 19:18:04 Listing swarm cluster nodes on VM [10.161.4.212]
2017/05/30 19:18:04 Creating volume [swarm_test_volume_1496171884] on VM [10.161.4.212]
2017/05/30 19:18:06 Confirming detached status for volume [swarm_test_volume_1496171884]
2017/05/30 19:18:06 Sleep for 2 seconds
2017/05/30 19:18:10 START: swarm_test.TestDockerSwarm
2017/05/30 19:18:10 Creating a docker service swarm_test_service_1496171884 on VM [10.161.4.212]
2017/05/30 19:18:12 Checking docker service [swarm_test_service_1496171884] status on VM [10.161.4.212]
2017/05/30 19:18:12 Listing docker services running on VM [10.161.4.212]
2017/05/30 19:18:12 Checking running containers for docker service [swarm_test_service_1496171884] on docker hosts: [10.161.4.212 10.161.28.209 10.161.27.126]
2017/05/30 19:18:13 Containers running on docker host [10.161.4.212]: swarm_test_service_1496171884.1.4jb1meifdtivsykhw2z65ybb2
2017/05/30 19:18:13 Confirming attached status for volume [swarm_test_volume_1496171884]
2017/05/30 19:18:15 Scaling 2 replicated services for swarm_test_service_1496171884 on VM [10.161.4.212]
2017/05/30 19:18:16 Checking docker service [swarm_test_service_1496171884] status on VM [10.161.4.212]
2017/05/30 19:18:16 Listing docker services running on VM [10.161.4.212]
2017/05/30 19:18:16 Checking running containers for docker service [swarm_test_service_1496171884] on docker hosts: [10.161.4.212 10.161.28.209 10.161.27.126]
2017/05/30 19:18:17 Containers running on docker host [10.161.4.212]: swarm_test_service_1496171884.1.4jb1meifdtivsykhw2z65ybb2
2017/05/30 19:18:17 Sleep for 5 seconds
2017/05/30 19:18:22 Containers running on docker host [10.161.4.212]: swarm_test_service_1496171884.1.4jb1meifdtivsykhw2z65ybb2
2017/05/30 19:18:22 Sleep for 5 seconds
2017/05/30 19:18:28 Containers running on docker host [10.161.4.212]: swarm_test_service_1496171884.1.4jb1meifdtivsykhw2z65ybb2
2017/05/30 19:18:28 Sleep for 5 seconds
2017/05/30 19:18:33 Containers running on docker host [10.161.4.212]: swarm_test_service_1496171884.1.4jb1meifdtivsykhw2z65ybb2
2017/05/30 19:18:33 Sleep for 5 seconds
2017/05/30 19:18:39 Containers running on docker host [10.161.4.212]: swarm_test_service_1496171884.1.4jb1meifdtivsykhw2z65ybb2
2017/05/30 19:18:39 Sleep for 5 seconds
2017/05/30 19:18:44 Containers running on docker host [10.161.4.212]: swarm_test_service_1496171884.1.4jb1meifdtivsykhw2z65ybb2
2017/05/30 19:18:44 Sleep for 5 seconds
2017/05/30 19:18:49 Containers running on docker host [10.161.4.212]: swarm_test_service_1496171884.1.4jb1meifdtivsykhw2z65ybb2
2017/05/30 19:18:49 Sleep for 5 seconds
2017/05/30 19:18:55 Containers running on docker host [10.161.4.212]: swarm_test_service_1496171884.1.4jb1meifdtivsykhw2z65ybb2
2017/05/30 19:18:55 Sleep for 5 seconds
2017/05/30 19:19:00 Containers running on docker host [10.161.4.212]: swarm_test_service_1496171884.1.4jb1meifdtivsykhw2z65ybb2
2017/05/30 19:19:00 Sleep for 5 seconds
2017/05/30 19:19:06 Containers running on docker host [10.161.4.212]: swarm_test_service_1496171884.1.4jb1meifdtivsykhw2z65ybb2
2017/05/30 19:19:06 Sleep for 5 seconds
2017/05/30 19:19:11 Containers running on docker host [10.161.4.212]: swarm_test_service_1496171884.1.4jb1meifdtivsykhw2z65ybb2
2017/05/30 19:19:11 Sleep for 5 seconds
2017/05/30 19:19:17 Containers running on docker host [10.161.4.212]: swarm_test_service_1496171884.2.ypfvob4ti7e8qczzv3qct41c2
swarm_test_service_1496171884.1.4jb1meifdtivsykhw2z65ybb2
2017/05/30 19:19:17 Stopping docker service swarm_test_service_1496171884.1.4jb1meifdtivsykhw2z65ybb2 on VM [10.161.4.212]
2017/05/30 19:19:28 Checking docker service [swarm_test_service_1496171884] status on VM [10.161.4.212]
2017/05/30 19:19:28 Listing docker services running on VM [10.161.4.212]
2017/05/30 19:19:28 Checking running containers for docker service [swarm_test_service_1496171884] on docker hosts: [10.161.4.212 10.161.28.209 10.161.27.126]
2017/05/30 19:19:29 Containers running on docker host [10.161.4.212]: swarm_test_service_1496171884.2.ypfvob4ti7e8qczzv3qct41c2
2017/05/30 19:19:29 Sleep for 5 seconds
2017/05/30 19:19:34 Containers running on docker host [10.161.4.212]: swarm_test_service_1496171884.1.n552uhd3gmeup01dwix1nqrid
swarm_test_service_1496171884.2.ypfvob4ti7e8qczzv3qct41c2
2017/05/30 19:19:34 Confirming attached status for volume [swarm_test_volume_1496171884]
2017/05/30 19:19:37 Destroying volume [swarm_test_volume_1496171884]
2017/05/30 19:19:37 Failed to invoke command [docker volume rm swarm_test_volume_1496171884]: exit status 1
2017/05/30 19:19:37 Removing docker service swarm_test_service_1496171884 on VM [10.161.4.212]
2017/05/30 19:19:38 Listing docker services running on VM [10.161.4.212]
2017/05/30 19:19:38 Failed to invoke command [docker service ps swarm_test_service_1496171884]: exit status 1
2017/05/30 19:19:38 Confirming detached status for volume [swarm_test_volume_1496171884]
2017/05/30 19:19:38 Sleep for 2 seconds
2017/05/30 19:19:41 Sleep for 2 seconds
2017/05/30 19:19:44 Sleep for 2 seconds
2017/05/30 19:19:47 Sleep for 2 seconds
2017/05/30 19:19:51 END: swarm_test.TestDockerSwarm
2017/05/30 19:19:51 Destroying volume [swarm_test_volume_1496171884]
2017/05/30 19:19:52 Finding VM name from IP Address [10.161.28.209]
2017/05/30 19:19:53 START: Test vmdkops service restart + kill vm process
2017/05/30 19:19:53 Creating volume [vmlistener_test_volume_1496171993] on VM [10.161.28.209]
2017/05/30 19:19:55 Attaching volume [vmlistener_test_volume_1496171993] on VM [10.161.28.209]
2017/05/30 19:19:57 Confirming attached status for volume [vmlistener_test_volume_1496171993]
2017/05/30 19:19:59 Restarting vmdkops service on ESX host [10.161.10.254]
2017/05/30 19:20:02 Confirming attached status for volume [vmlistener_test_volume_1496171993]
2017/05/30 19:20:06 Retrieving VM power state for [VM1.2]
2017/05/30 19:20:07 VM[VM1.2]'s current power state is [poweredOn]
2017/05/30 19:20:07 Retrieving World ID for VM [VM1.2] from ESX [10.161.10.254]
2017/05/30 19:20:07 VM's process ID is: 6134808
2017/05/30 19:20:07 Killing VM from ESX [10.161.10.254]
2017/05/30 19:20:08 Confirming [poweredOff] power status for vm [VM1.2]
2017/05/30 19:20:08 Sleep for 3 seconds
2017/05/30 19:20:11 Retrieving VM power state for [VM1.2]
2017/05/30 19:20:14 Powering on VM [VM1.2]
2017/05/30 19:20:16 Confirming [poweredOn] power status for vm [VM1.2]
2017/05/30 19:20:16 Sleep for 3 seconds
2017/05/30 19:20:19 Retrieving VM power state for [VM1.2]
2017/05/30 19:20:20 Confirming detached status for volume [vmlistener_test_volume_1496171993]
2017/05/30 19:20:20 Sleep for 2 seconds
2017/05/30 19:20:31 Failed to invoke command [docker volume inspect  --format '{{index .Status.status}}' vmlistener_test_volume_1496171993]: exit status 255
2017/05/30 19:20:31 Sleep for 2 seconds
2017/05/30 19:20:33 Failed to invoke command [docker volume inspect  --format '{{index .Status.status}}' vmlistener_test_volume_1496171993]: exit status 255
2017/05/30 19:20:33 Sleep for 2 seconds
2017/05/30 19:20:43 END: Test vmdkops service restart + kill vm process
2017/05/30 19:20:43 Checking container [vmlistener_test_container_1496171993] presence on VM [10.161.28.209]
2017/05/30 19:20:44 Destroying volume [vmlistener_test_volume_1496171993]
2017/05/30 19:20:44 Failed to invoke command [docker volume rm vmlistener_test_volume_1496171993]: exit status 1

----------------------------------------------------------------------
FAIL: vmlistener_test.go:61: VMListenerTestParams.TearDownTest

vmlistener_test.go:67:
    c.Assert(err, IsNil, Commentf(out))
... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc4201673a0), Stderr:[]uint8(nil)} ("exit status 1")
... Error response from daemon: unable to remove volume: remove vmlistener_test_volume_1496171993: VolumeDriver.Remove: Remove failure - volume is still mounted.  volume=vmlistener_test_volume_1496171993, refcount=1


----------------------------------------------------------------------
PANIC: vmlistener_test.go:80: VMListenerTestParams.TestKillVM

... Panic: Fixture has panicked (see related PANIC)
OOPS: 8 passed, 1 FAILED, 1 MISSED
--- FAIL: Test (269.22s)
=== RUN   TestVolumeProperties
2017/05/30 19:20:44 running end_to_end tests - current time:  2017-05-30 19:20:44.834216839 +0000 UTC
2017/05/30 19:20:44 Running test on VM -  10.161.4.212
2017/05/30 19:20:44 Creating a volume of Format Type -  thin
2017/05/30 19:20:48 
 successfully created a  volume --  volumeprop_test_volume_1496172044
2017/05/30 19:20:48 Verifying volume properties like size, disk-format and attached-to-vm fields at vm and esx for volume -  volumeprop_test_volume_1496172044
2017/05/30 19:20:56 Finished verifying volume properties like size, disk-format and attached-to-vm fields at vm and esx for volume -  volumeprop_test_volume_1496172044
2017/05/30 19:21:08 Creating a volume of Format Type -  zeroedthick
2017/05/30 19:21:10 
 successfully created a  volume --  volumeprop_test_volume_1496172068
2017/05/30 19:21:10 Verifying volume properties like size, disk-format and attached-to-vm fields at vm and esx for volume -  volumeprop_test_volume_1496172068
2017/05/30 19:21:17 Finished verifying volume properties like size, disk-format and attached-to-vm fields at vm and esx for volume -  volumeprop_test_volume_1496172068
2017/05/30 19:21:28 Creating a volume of Format Type -  eagerzeroedthick
2017/05/30 19:21:31 
 successfully created a  volume --  volumeprop_test_volume_1496172088
2017/05/30 19:21:31 Verifying volume properties like size, disk-format and attached-to-vm fields at vm and esx for volume -  volumeprop_test_volume_1496172088
2017/05/30 19:21:39 Finished verifying volume properties like size, disk-format and attached-to-vm fields at vm and esx for volume -  volumeprop_test_volume_1496172088
--- PASS: TestVolumeProperties (65.71s)
FAIL
2017/05/30 19:21:52 -----Clean-up finished - current time:  2017-05-30 19:21:52.060603163 +0000 UTC
exit status 1
FAIL	github.com/vmware/docker-volume-vsphere/tests/e2e	336.449s
make: *** [e2e-dkrVolDriver-test] Error 1
make: Leaving directory `/go/src/github.com/vmware/docker-volume-vsphere/vmdk_plugin'
make[1]: *** [e2e-test] Error 2
make: *** [e2e-test] Error 2

**Comment by shuklanirdesh82** [comment]
Thanks @shaominchen #1306 takes care of this one.

**Comment by shaominchen** [comment]
Reopen this issue as it still happens intermittently. We have missed one step in previous fix - to check the volume detached status. This step is REQUIRED because stopping or removing a container will NOT release the volume immediately, since it takes some time for the container to stop or exit. We must use a poller to check and verify the volume status is detached before removing it. Please refer to basic_test as an example: https://github.com/vmware/docker-volume-vsphere/blob/master/tests/e2e/basic_test.go#L84

**Comment by shuklanirdesh82** [comment]
> We have missed one step in previous fix - to check the volume detached status. This step is REQUIRED because stopping or removing a container will NOT release the volume immediately, since it takes some time for the container to stop or exit. We must use a poller to check and verify the volume status is detached before removing it. Please refer to basic_test as an example: https://github.com/vmware/docker-volume-vsphere/blob/master/tests/e2e/basic_test.go#L84

1. Test case step https://github.com/vmware/docker-volume-vsphere/blob/master/tests/e2e/vmlistener_test.go#L126 does make sure `detached` state before invoking `docker volume rm`.

2. Testcase does uses polling mechanism to check upon power state and detach/attach state. Only gap here is about `refCount` thread state (see https://github.com/vmware/docker-volume-vsphere/pull/1300#discussion_r118602099)

3. At present there is no way to know about refCount operation is in progress or completed. (`vDVS plugin refcount thread is running in background and no hook to hang on reliably`)

As discussed offline we need to improve plugin error message that returns `refcount=-1` when the operation is in-progress ... accordingly change is needed for `docker volume rm ` util.

There is nothing left for the issue so closing this issue.

**Comment by shaominchen** [comment]
@shuklanirdesh82 I have filed issue #1348 for vmdk plugin issue. However, after the plugin issue gets fixed, we still need to fix this test case accordingly to check the "volume rm" error messages. So please file a new issue (since the current one is closed) and mark it depending on issue #1348.

**Comment by shuklanirdesh82** [comment]
#1349 is opened to track the improvement.
